pandas/core/indexes/base.py

```diff
@@ -1196,15 +1194,26 @@ def _validate_names(self, name=None, names=None, deep: bool = False):
         if names is not None and name is not None:
             raise TypeError("Can only provide one of `names` and `name`")
         elif names is None and name is None:
-            return deepcopy(self.names) if deep else self.names
+            new_names = deepcopy(self.names) if deep else self.names
         elif names is not None:
             if not is_list_like(names):
                 raise TypeError("Must pass list-like as `names`.")
-            return names
+            new_names = names
+        elif not is_list_like(name):
+            new_names = [name]
         else:
-            if not is_list_like(name):
-                return [name]
-            return name
+            new_names = name
+
+        if len(new_names) != len(self.names):
+            raise ValueError(
+                f"Length of new names must be {len(self.names)}, got {len(new_names)}"
+            )
+        # All items in 'new_names' need to be hashable
+        for new_name in new_names:
+            if not is_hashable(new_name):
+                raise TypeError(f"{type(self).__name__}.name must be a hashable type")
```

Comment:
are there tests which catch these cases?

Reply:
hmm, we check hashability elsewhere do we not?

Reply:
In master it’s checked in .set_names. As this PR is done to avoid that method in .copy, hashabiliy now needs to be tested in  _validate_names also.

Reply:
can you make a common method to check?